seqr/views/utils/dataset_utils.py

```diff
@@ -298,67 +320,103 @@ def _parse_rna_row(row, column_map, required_column_map, missing_required_fields
+):
 
     sample_id_to_individual_id_mapping = {}
     if mapping_file:
         sample_id_to_individual_id_mapping = load_mapping_file_content(mapping_file)
 
-    samples_by_id = defaultdict(dict)
+    samples_by_guid = defaultdict(dict)
     f = file_iter(file_path, user=user)
-    header = _parse_tsv_row(next(f))
+    parsed_f = parse_file(file_path.replace('.gz', ''), f, iter_file=True)
+    header = next(parsed_f)
     required_column_map = _validate_rna_header(header, column_map)
 
-    sample_id_to_tissue_type = {}
-    samples_with_conflict_tissues = defaultdict(set)
-    errors = []
+    loaded_samples = set()
+    unmatched_samples = set()
     missing_required_fields = defaultdict(list)
     gene_ids = set()
+    current_sample = None
     for line in tqdm(f, unit=' rows'):
         row = dict(zip(header, _parse_tsv_row(line)))
         for sample_id, row_dict in _parse_rna_row(
                 row, column_map, required_column_map, missing_required_fields, allow_missing_gene, **kwargs):
             tissue_type = TISSUE_TYPE_MAP[row[TISSUE_COL]]
             project = row[PROJECT_COL]
-            if (sample_id, project) in sample_id_to_tissue_type:
-                prev_tissue_type = sample_id_to_tissue_type[(sample_id, project)]
-                if tissue_type != prev_tissue_type:
-                    samples_with_conflict_tissues[(sample_id, project)].update({prev_tissue_type, tissue_type})
-                    continue
+            sample_key = (sample_id, project, tissue_type)
+
+            if sample_key in potential_loaded_samples:
+                loaded_samples.add(sample_key)
+                continue
 
-            sample_id_to_tissue_type[(sample_id, project)] = tissue_type
+            if row.get(INDIV_ID_COL) and sample_id not in sample_id_to_individual_id_mapping:
+                sample_id_to_individual_id_mapping[sample_id] = row[INDIV_ID_COL]
 
             gene_id = row_dict[GENE_ID_COL]
             if gene_id:
                 gene_ids.add(gene_id)
 
-            if get_unique_key:
-                gene_or_unique_id = get_unique_key(row_dict)
-            else:
-                gene_or_unique_id = gene_id
-            existing_data = samples_by_id[(sample_id, project)].get(gene_or_unique_id)
+            sample_guid = get_matched_sample(sample_key, unmatched_samples, sample_id_to_individual_id_mapping)
+
+            if missing_required_fields or (unmatched_samples and not ignore_extra_samples) or (sample_key in unmatched_samples):
+                # If there are definite errors, do not process/save data, just continue to check for additional errors
+                continue
+
+            if current_sample != sample_guid:
+                if len(samples_by_guid[current_sample]) > 5000:
+                    save_sample_data(current_sample, samples_by_guid[current_sample])
+                    del samples_by_guid[current_sample]
+                current_sample = sample_guid
```

Comment:
I don't follow what's happening here - why are you checking `if len(samples_by_guid[current_sample]) > 5000`? should `5000` be a constant?

Reply:
It should be a constant. The idea here is we do not want to be holding so much in memory, so if a substantive amount of data has been parsed for a sample we should save that data and continue. I'll add a comment to make it clearer